packages/js/components/src/image-gallery/image-gallery.tsx

```diff
@@ -17,17 +24,103 @@ export const ImageGallery: React.FC< ImageGalleryProps > = ( {
 	children,
 	columns = 4,
 }: ImageGalleryProps ) => {
+	const [ toolBarItem, setToolBarItem ] = useState<
+		JSX.Element | null | undefined
+	>( null );
+	const [ orderedChildren, setOrderedChildren ] = useState< JSX.Element[] >(
+		[]
+	);
+
+	useEffect( () => {
+		if ( ! children ) {
+			return;
+		}
+		setOrderedChildren(
+			Array.isArray( children ) ? children : [ children ]
+		);
+	}, [ children ] );
+
+	const moveItem = ( fromIndex: number, toIndex: number ) => {
+		setOrderedChildren(
+			moveIndex< JSX.Element >( fromIndex, toIndex, orderedChildren )
+		);
+	};
+
+	const removeItem = ( removeIndex: number ) => {
+		setOrderedChildren(
+			orderedChildren.filter( ( _, index ) => index !== removeIndex )
+		);
+	};
+
+	const setAsCoverImage = ( coverIndex: number ) => {
+		moveItem( coverIndex, 0 );
+		setToolBarItem( null );
+	};
+
 	return (
 		<div
 			className="woocommerce-image-gallery"
 			style={ {
 				gridTemplateColumns: 'min-content '.repeat( columns ),
 			} }
 		>
-			<Sortable isHorizontal>
-				{ Children.map( children, ( child ) => (
-					<SortableHandle>{ child }</SortableHandle>
-				) ) }
+			<Sortable
+				isHorizontal
+				onOrderChange={ ( items ) => {
+					setOrderedChildren( items );
+				} }
+				onDragStart={ () => setToolBarItem( null ) }
+			>
+				{ Children.map( orderedChildren, ( child, childIndex ) => {
+					const isCoverImage = childIndex === 0;
+
+					return (
+						<div
+							className={ `woocommerce-image-gallery__item-wrapper ${
+								isCoverImage ? 'not-sortable ' : ''
+							}` }
+							//TODO: add correct keyboard handler
+							onKeyPress={ () => {} }
+							tabIndex={ 0 }
+							role="button"
+							onClick={ ( event ) => {
+								if (
+									( event.target as Element ).closest(
+										'.woocommerce-image-gallery__toolbar'
+									) !== null
+								) {
+									return;
+								}
+								setToolBarItem(
+									Boolean( child ) && toolBarItem === child
+										? null
+										: child
+								);
+							} }
+						>
+							{ toolBarItem === child && (
+								<ImageGalleryToolbar
+									childIndex={ childIndex }
+									lastChild={
+										childIndex ===
+										orderedChildren.length - 1
+									}
+									moveItem={ moveItem }
+									removeItem={ removeItem }
+									setAsCoverImage={ setAsCoverImage }
+								/>
+							) }
+
+							{ isCoverImage ? (
+								cloneElement( child, {
+									isCover: true,
+								} )
+							) : (
+								<SortableHandle>{ child }</SortableHandle>
```

Comment:
I think what's happening here is that when you reorder, it's passing back the children exactly as they are.  In your case, it looks like the children might be:

```
<ImageGalleryItem />
<SortableHandle><ImageGalleryItem /></SortableHandle>
<SortableHandle><ImageGalleryItem /></SortableHandle>
<SortableHandle><ImageGalleryItem /></SortableHandle>
...
```

On the next render after you've updated state with the reordered items, when you check the item, even though `isCoverImage` is `true`, the first item would look like this:

```
<SortableHandle isCover><ImageGalleryItem /></SortableHandle>
```

I have a few recommendations in order of what I would consider highest priority:

1. Is it possible just to make the first item `isCover`?
3. Instead of tracking the whole child as the `coverImage`, I think we should just track a `key` or ID of some sort
4. `SortableHandle` is probably best moved inside the `ImageGalleryItem` to make it a little easier for you to pass props around
5. Optionally, we could pass props from `SortableHandle` down to any children it has.

Reply:
Thanks for this @joshuatf , the issue with the nested children was my primary issue since I wasn't looking in the right places. Appreciate it!

> Is it possible just to make the first item isCover?

This has been my approach, so I'm unsure if you're talking about something different. The item under the first index (`0`) is always set as the cover. Let me know if I'm misunderstanding something.

> Instead of tracking the whole child as the coverImage, I think we should just track a key or ID of some sort \

This had been my intention but I had mystery keys appearing in unpredictable ways, and only too late for me to utilize them. I now believe this was stemming from the `Children.map()` function(s) which automaticly add them, and I've migrated this to set my own keys on the base children as they're initialized and tracking them with those. Seems to work! 👍🏻 

> SortableHandle is probably best moved inside the ImageGalleryItem to make it a little easier for you to pass props around

Yup, good call and I've moved all wrapping functions within `ImageGalleryItem`. 

> Optionally, we could pass props from SortableHandle down to any children it has.

Doesn't seem necessary at the moment, but glad to throw this in if desired.

Reply:
>This has been my approach, so I'm unsure if you're talking about something different. The item under the first index (0) is always set as the cover. Let me know if I'm misunderstanding something.

I don't think you'll need to track the `coverImage` in state in this case.  You can just grab the first item from the list and pass the `isCover` prop to it, right?

Reply:
Scratch that.  Just looked at the latest iteration and it does not track cover image in state.